manifest.go

```diff
@@ -0,0 +1,655 @@
+	return fetchManifestEntries(m, fs, discardDeleted)
+}
+
+func fetchManifestEntries(m ManifestFile, fs iceio.IO, discardDeleted bool) ([]ManifestEntry, error) {
+	f, err := fs.Open(m.FilePath())
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	dec, err := ocf.NewDecoder(f)
+	if err != nil {
+		return nil, err
+	}
+
+	metadata := dec.Metadata()
+	isVer1 := true
+	if string(metadata["format-version"]) == "2" {
+		isVer1 = false
+	}
+
+	results := make([]ManifestEntry, 0)
+	for dec.HasNext() {
+		var tmp ManifestEntry
+		if isVer1 {
+			tmp = &manifestEntryV1{}
+		} else {
+			tmp = &manifestEntryV2{}
+		}
+
+		if err := dec.Decode(tmp); err != nil {
+			return nil, err
+		}
+
+		if !discardDeleted || tmp.Status() != EntryStatusDELETED {
+			tmp.inheritSeqNum(m)
+			results = append(results, tmp)
+		}
+	}
+
+	return results, dec.Error()
+}
+
+// ManifestFile is the interface which covers both V1 and V2 manifest files.
+type ManifestFile interface {
+	// Version returns the version number of this manifest file.
+	// It should be 1 or 2.
+	Version() int
+	// FilePath is the location URI of this manifest file.
+	FilePath() string
+	// Length is the length in bytes of the manifest file.
+	Length() int64
+	// PartitionID is the ID of the partition spec used to write
+	// this manifest. It must be listed in the table metadata
+	// partition-specs.
+	PartitionID() int32
+	// ManifestContent is the type of files tracked by this manifest,
+	// either data or delete files. All v1 manifests track data files.
+	ManifestContent() ManifestContent
+	// SnapshotID is the ID of the snapshot where this manifest file
+	// was added.
+	SnapshotID() int64
+	// AddedDataFiles returns the number of entries in the manifest that
+	// have the status of EntryStatusADDED.
+	AddedDataFiles() int32
+	// ExistingDataFiles returns the number of entries in the manifest
+	// which have the status of EntryStatusEXISTING.
+	ExistingDataFiles() int32
+	// DeletedDataFiles returns the number of entries in the manifest
+	// which have the status of EntryStatusDELETED.
+	DeletedDataFiles() int32
+	// AddedRows returns the number of rows in all files of the manifest
+	// that have status EntryStatusADDED.
+	AddedRows() int64
+	// ExistingRows returns the number of rows in all files of the manifest
+	// which have status EntryStatusEXISTING.
+	ExistingRows() int64
+	// DeletedRows returns the number of rows in all files of the manifest
+	// which have status EntryStatusDELETED.
+	DeletedRows() int64
+	// SequenceNum returns the sequence number when this manifest was
+	// added to the table. Will be 0 for v1 manifest lists.
+	SequenceNum() int64
+	// MinSequenceNum is the minimum data sequence number of all live data
+	// or delete files in the manifest. Will be 0 for v1 manifest lists.
+	MinSequenceNum() int64
+	// Metadata returns implementation-specific key metadata for encryption
+	// if it exists in the manifest list.
+	Metadata() []byte
```

Comment:
I'd rather name this `KeyMetadata()` to have a clear distinction later between this and table metadata